gum/backend-darwin/gumdarwinmodule.c

```diff
@@ -1268,6 +1298,51 @@ gum_emit_section_term_pointers (const GumDarwinSectionDetails * details,
   return TRUE;
 }
 
+void
+gum_darwin_module_enumerate_dependencies (GumDarwinModule * self,
+                                          GumDarwinFoundDependencyFunc func,
+                                          gpointer user_data)
+{
+  struct mach_header * header;
+  gconstpointer command;
+  gsize command_index;
+
+  if (!gum_darwin_module_ensure_image_loaded (self, NULL))
+    return;
+
+  header = (struct mach_header *) self->image->data;
+  if (header->magic == MH_MAGIC)
+    command = self->image->data + sizeof (struct mach_header);
+  else
+    command = self->image->data + sizeof (struct mach_header_64);
+  for (command_index = 0; command_index != header->ncmds; command_index++)
+  {
+    const struct load_command * lc = (struct load_command *) command;
+
+    switch (lc->cmd)
+    {
+      case LC_LOAD_DYLIB:
+      case LC_LOAD_WEAK_DYLIB:
+      case LC_REEXPORT_DYLIB:
+      case LC_LOAD_UPWARD_DYLIB:
+      {
+        const struct dylib_command * dc = command;
+        const gchar * name;
+
+        name = command + dc->dylib.name.offset;
+        if (!func (name, user_data))
+          return;
+
+        break;
+      }
+      default:
+        break;
+    }
+
+    command += lc->cmdsize;
+  }
```

Comment:
Wondering if we should roll this into the load step, so we can populate `dependencies` like we normally do.

Reply:
yeah right, that was my first thought too since we're already storing dependencies in `_take_image`. my first implementation was:
```
  const gchar * dep;
  gint index = 0;

  if (!gum_darwin_module_ensure_image_loaded (self, NULL))
    return;

  while ((dep = g_ptr_array_index (self->dependencies, index++)) != NULL)
  {
    if (!func (dep, user_data))
      return;
  }
```
but this seems to cause some kind of corruption and everything ends up hanging and i have to restart SpringBoard. i'm looking into that - not sure why it's behaving like this since the pointers should be stable as long as the image is there.

Reply:
It's because `g_ptr_array_index()` intentionally doesn't do any bounds-checking, so you are meant to use `->len`.

Reply:
ahh "nice" !!